src/H5Odtype.c

```diff
@@ -1099,30 +1280,47 @@ H5O_dtype_encode_helper(const H5F_t *f, uint8_t **pp, const H5T_t *dt)
 --------------------------------------------------------------------------*/
 static void *
 H5O_dtype_decode(H5F_t *f, H5O_t H5_ATTR_UNUSED *open_oh, unsigned H5_ATTR_UNUSED mesg_flags,
-                 unsigned *ioflags /*in,out*/, size_t H5_ATTR_UNUSED p_size, const uint8_t *p)
+                 unsigned *ioflags /*in,out*/, size_t p_size, const uint8_t *p)
 {
-    H5T_t *dt        = NULL;
-    void  *ret_value = NULL; /* Return value */
+    hbool_t        skip;
+    H5T_t         *dt        = NULL;
+    const uint8_t *p_end     = p + p_size - 1;
+    void          *ret_value = NULL;
 
     FUNC_ENTER_NOAPI_NOINIT
 
-    /* check args */
+    HDassert(f);
     HDassert(p);
 
     /* Allocate datatype message */
     if (NULL == (dt = H5T__alloc()))
         HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, NULL, "memory allocation failed")
 
+    /* If we are decoding a buffer from H5Tdecode(), we won't have the size
+     * of the buffer and bounds checking will be impossible. In this case,
+     * the library will have set p_size to SIZE_MAX and we can use that
+     * as a signal to skip bounds checking.
+     */
+    skip = (p_size == SIZE_MAX ? TRUE : FALSE);
+
     /* Perform actual decode of message */
-    if (H5O_dtype_decode_helper(f, ioflags, &p, dt) < 0)
+    if (H5O_dtype_decode_helper(f, ioflags, &p, dt, skip, p_end) < 0)
         HGOTO_ERROR(H5E_DATATYPE, H5E_CANTDECODE, NULL, "can't decode type")
 
     /* Set return value */
     ret_value = dt;
 
 done:
+    /* Cleanup on error */
+    if (!ret_value) {
+        if (dt != NULL) {
+            if (dt->shared != NULL)
+                dt->shared = H5FL_FREE(H5T_shared_t, dt->shared);
```

Comment:
Doing this instead of `H5T_close_real(dt)` avoids problems in cases where `dt->shared` is `NULL`.

Reply:
Is the develop branch not failing in the same way? Since it uses `H5T_close_real(dt)` we should figure out the difference between the two branches so that they are kept in sync with each other.

Reply:
This was just an issue with this branch.  Upon further examination, the actual root issue was duplication of freeing logic introduced during the merge; I've set this back to `H5T_close_real`.